ml-agents/mlagents/trainers/lesson_controller.py

```diff
@@ -0,0 +1,122 @@
+import os
+import yaml
+import math
+
+from .exception import LessonControllerError
+
+import logging
+
+logger = logging.getLogger("mlagents.trainers")
+
+
+class LessonController(object):
+    def __init__(self, location):
+        """
+        Initializes a Curriculum object.
+        :param location: Path to yaml file defining reset configuration
+        """
+        self.measure = None
+        try:
+            with open(location) as data_file:
+                self.data = yaml.load(data_file)
+        except IOError:
+            raise LessonControllerError("The file {0} could not be found.".format(location))
+        except UnicodeDecodeError:
+            raise LessonControllerError("There was an error decoding {}".format(location))
+        self.smoothing_value = 0
+        self.check_keys(location)
+        self.measure = self.data["measure"]
+        self.thresholds = self.data["thresholds"]
+        self.min_lesson_length = self.data["min_lesson_length"]
+        self.max_lesson_num = len(self.thresholds)
+        self._lesson_num = 0
+
+
+    def check_keys(self, location):
+        for key in [
+            "measure",
+            "thresholds",
+            "min_lesson_length",
+            "signal_smoothing",
+        ]:
+            if key not in self.data:
+                raise LessonControllerError(
+                    "{0} does not contain a " "{1} field.".format(location, key)
+                )
+
+
+    @property
+    def lesson_num(self):
+        return self._lesson_num
+
+    @lesson_num.setter
+    def lesson_num(self, lesson_num):
+        self._lesson_num = max(0, min(lesson_num, self.max_lesson_num))
+
+    def _lesson_ready_to_increment(self, brain_name, reward_buff_size):
+        """Determines whether the curriculum of a specified brain is ready
+        to attempt an increment.
+
+        Args:
+            brain_name (str): The name of the brain whose curriculum will be
+                checked for readiness.
+            reward_buff_size (int): The size of the reward buffer of the trainer
+                that corresponds to the specified brain.
+
+        Returns:
+            Whether the curriculum of the specified brain should attempt to
+            increment its lesson.
+        """
+        return (reward_buff_size >= self.min_lesson_length)
+
+    def change_lesson(self, measure_val):
```

Comment:
nit: `should_change_lesson`?

Reply:
Hmm, maybe not. Ideally I'd like to separate the decision from the action, but that's harder with the smoothing. Need to think about it.